### PR TITLE
Better handling of uncaught exceptions in main and user threads

### LINKED ISSUE #3418: Can we provide a friendlier default handler for unhandled exceptions?
Seems like if you have an exception thrown while rendering another exception, you get a pretty scary C++ ABI error.

Consider this:

```scala
case class MyErr(msg: String) extends Exception:
  override def toString() = msg(0) + "!"

@main def hello = 
  throw MyErr("")
```

See the difference between JVM and Native:

```
$ scli run .
Exception in thread "main"
Exception: java.lang.StringIndexOutOfBoundsException thrown from the UncaughtExceptionHandler in thread "main"

$ scli run . --native
libc++abi: terminating due to uncaught exception of type scalanative::ExceptionWrapper: std::exception
fish: Job 1, 'scala-cli $argv' terminated by signal SIGABRT (Abort)
```

Note how there's at least _some_ clue in JVM version as to what the original exception was

### PR DESCRIPTION
**Pull request by WojciechMazur**
Fixes #3418

### CODE REVIEW

**Comment by LeeTibbert** [comment]
Suggested change in title: s/uncought/uncaught/

File: javalib/src/main/scala/java/lang/Thread.scala [hunk 1]
```diff
@@ -353,6 +353,22 @@ class Thread private[lang] (
 object Thread {
   trait UncaughtExceptionHandler {
     def uncaughtException(t: Thread, e: Throwable): Unit
+
+    // ScalaNativeSpecific exception handler, used when executing user-provied handlers, in main
+    // Duplicate in NativeThread.threadEntryPoint
+    private[java] final def uncaughtExceptionInternal(
```
Comment (tanishiking) [inline javalib/src/main/scala/java/lang/Thread.scala:359]: [note] We come here when scala-native app fails with uncaught exception

File: javalib/src/main/scala/java/lang/Thread.scala [hunk 2]
```diff
@@ -353,6 +353,22 @@ class Thread private[lang] (
...
+        t: Thread,
+        e: Throwable
+    ): Unit = {
+      try uncaughtException(t, e)
+      catch {
```
Comment (tanishiking) [inline javalib/src/main/scala/java/lang/Thread.scala:364]: [note]
here we catch the exception thrown while rendering another exception

File: javalib/src/main/scala/java/lang/Thread.scala [hunk 3]
```diff
@@ -548,10 +564,10 @@ object Thread {
   def getAllStackTraces(): java.util.Map[Thread, Array[StackTraceElement]] =
     throw new UnsupportedOperationException()
 
-  private var defaultExceptionHandler: UncaughtExceptionHandler = _
+  @volatile private var defaultExceptionHandler: UncaughtExceptionHandler = _
   def getDefaultUncaughtExceptionHandler(): UncaughtExceptionHandler =
     defaultExceptionHandler
-  def setDefaultUncaughtHandler(eh: UncaughtExceptionHandler): Unit =
+  def setDefaultUncaughtExceptionHandler(eh: UncaughtExceptionHandler): Unit =
```
Comment (tanishiking) [inline javalib/src/main/scala/java/lang/Thread.scala:570]: Maybe a silly question, who calls `setDefaultUncaughtExceptionHandler`? I couldn't find the caller of this method in this project, and `ThreadGroup#uncaughtException` always goes this branch https://github.com/natsukagami/scala-native/blob/ac2460c1fa20081ca82a18cc564e7686136a17a9/javalib/src/main/scala/java/lang/ThreadGroup.scala#L222-L225 ?

**Review by tanishiking — APPROVED** [review]
It looks good to me, now we will have

```
Exception in thread "main"
Exception: java.lang.StringIndexOutOfBoundsException thrown from the UncaughtExceptionHandler in thread "main"
... (stacktrace)
```

when exception is thrown while rendering another exception :+1:

I left one question about `setDefaultUncaughtExceptionHandler`.

**Review by tanishiking — CHANGES_REQUESTED** [review]
Review changes approve -> request changes, just because CI is failing :)

**Comment by JD557** [comment]
> Suggested change in title: s/uncought/uncaught/

Not just in the title, some places in the code also mention `uncought`

**Pull request merged**